### DISCUSSION TITLE
Supercharging USB Speed on Rock Pi E with dwc3!

### DISCUSSION THREAD

**Discussion by dynm** [marcone/teslausb#788] [Show and tell]
Hello everyone! 🌟

I've been experimenting with my Rock Pi E's USB3 and I've managed to resolve the USB low speed issue to Tesla's satisfaction! Now, I'm enjoying the brisk pace of USB 3. 🚀

Here's what I did:
1. Updated to the latest Armbian version.
2. Applied a dt overlay for dwc3 ([see the details here](https://gist.github.com/dynm/d493e2879abdb1473baa8280ff9575dd)) to avoid the "Enable of device-initiated U2 failed." and enable UASP device recognition.  Save this file and run this as root`armbian-add-overlay usb3-peripheral.dts`
3. Utilized a male-to-male USB3 cable and a jump wire to connect the USB3.0 port's VCC to the SBC's VCC, bypassing the USB3 port's power management, and ensuring a smooth power-on for the Rock Pi E.
<img width="610" alt="image" src="https://github.com/marcone/teslausb/assets/11208049/eb11ee90-3524-4e55-ba31-c7c7399b728c">

====== summary ======
hardware: Radxa ROCK Pi E
OS: Armbian 23.8.3 bookworm
ERROR: no archive method specified!
lun0 connected, from file /backingfiles/cam_disk.bin
lun1 connected, from file /backingfiles/music_disk.bin
2 snapshots mounted
Gadget state: configured

**Comment by marcone** [comment]
It sounds like you're saying the Rock Pi E supports USB3 OTG, but that's not what the specs say. If this is true, then that potentially opens up a bunch of other relatively cheap boards.
I like your use of a jumper wire to avoid having to use a separate USB cable for power.

**Reply by dynm** [reply]
About six months ago, I discovered that the rk3328 on RockPi E utilizes a dwc3 USB controller, indicating support for USB3.0 DRD. 
Upon attempting to apply a dt overlay to set the USB3 to peripheral mode, I realized it could only be recognized by certain USB Hosts. 
Interestingly, while my PC's front panel could identify it, other ports, and Tesla, could not. 
This led me to speculate that there might be an issue with the rk3328's USB3 controller, hence it wasn't highlighted as a feature. However, recently, I decided to revisit this by pulling out the RockPi E and delving into why it wasn't recognized by my PC's front panel USB ports. The dmesg logs provided some insights:
```
[7237970.907416] usb-storage 2-8:1.0: USB Mass Storage device detected
[7237970.907701] scsi host8: usb-storage 2-8:1.0
[7237971.231342] usb 2-8: USB disconnect, device number 127
[7237971.547492] usb 2-8: new SuperSpeed USB device number 3 using xhci_hcd
[7237971.568956] usb 2-8: New USB device found, idVendor=1d6b, idProduct=0104, bcdDevice= 1.00
[7237971.568959] usb 2-8: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[7237971.568960] usb 2-8: Product: TeslaUSB Composite Gadget
[7237971.568961] usb 2-8: Manufacturer: TeslaUSB
[7237971.568962] usb 2-8: SerialNumber: TeslaUSB-
[7237971.577585] usb 2-8: Enable of device-initiated U2 failed.
[7237976.795317] xhci_hcd 0000:00:14.0: WARN Set TR Deq Ptr cmd failed due to incorrect slot or ep state.
```
Then I ask to ChatGPT why it happens, and then it told me it may caused by driver's U1/U2 low power mode initialization, so I send some 
 driver's manual links to ChatGPT, finally, it gives me the correct driver quirk options.
https://www.kernel.org/doc/Documentation/devicetree/bindings/usb/dwc3.txt
https://github.com/torvalds/linux/commit/44dd5e2106dc2fd01697b539085818d1d1c58df0

**Reply by marcone** [reply]
That's some excellent detective work!

**Comment by THX723** [comment]
Very cool. How stable is the RockPi platform? The other [thread](https://github.com/marcone/teslausb/discussions/778) seems to imply it's still not quite ready for prime time.

**Reply by dynm** [reply]
I've been using RockPi E for a week now and haven't encountered any issues.

**Reply by THX723** [reply]
That's really awesome. What kind of archive performance are you getting over 5Ghz WiFi (Mbit/s)?

**Reply by dynm** [reply]
`rsync --stats -rP -L /mutable/TeslaCam/SavedClips/2023-10-19_08-15-45 root@192.168.0.100:/mnt/Main/TeslaUSB/SavedClips/`
```
Number of files: 47 (reg: 46, dir: 1)
Number of created files: 46 (reg: 46)
Number of deleted files: 0
Number of regular files transferred: 46
Total file size: 1,664,398,877 bytes
Total transferred file size: 1,664,398,877 bytes
Literal data: 1,664,398,877 bytes
Matched data: 0 bytes
File list size: 0
File list generation time: 0.001 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 1,664,808,480
Total bytes received: 891

sent 1,664,808,480 bytes  received 891 bytes  19,026,392.81 bytes/sec
total size is 1,664,398,877  speedup is 1.00
```
Tested with rsync over ssh, abount 19MB/s

**Reply by THX723** [reply]
152Mbps! That's fantastic. I average only 85Mbps with my RPi 4B. If indeed the E is rock solid, then there's no reason to go with anything else.

**Reply by dynm** [reply]
I experienced three flash card failures while using the OpiZero2. In searching for a solution, I discovered the RockPi E supports an eMMC module. However, its high cost led me to the OrangePi’s eMMC module, which is significantly more cheaper. After verifying their pin-to-pin compatibility, I purchased the OrangePi module and installed the OS on it. Now, the OS operates from the eMMC module, and I use the flash card as an external drive. This setup means if the flash card fails again, I can simply replace it and restore the partition table, avoiding the need for a new OS and TeslaUSB installation.

**Comment by forgottenruntime** [comment]
Nicely Done :-) . i was also looking for that board. but end up buying a Radxa Zero . it has USB2 OTG and one USB3 host . will it be possible to use your trick with this board? use USB3 OTG ?

**Reply by dynm** [reply]
It might not be possible to use the same trick with the Radxa Zero. This board uses the S905Y2 chipset, where the USB3 is provided by a USB 3.0/PCIe 2.0 Combo PHY. It seems like it doesn't support a 'dual role' for USB to act as a peripheral.

**Reply by ravibowman** [reply]
Any chance that you will make a guide on how you did this? I have been trying to get this to work on a RockPi 4C+ which does have USB3 OTG but for some reason, the implementation is not working even after doing everything needed to get OTG up.

**Reply by dynm** [reply]
I don't have a RockPi 4C+ right now. Have you tried "Part two" [in this page](https://wiki.radxa.com/Rockpi4/dev/usb-host-and-otg-switch-on-rock-pi-4) ?

**Reply by ravibowman** [reply]
Thanks for responding!

Those instructions no longer apply. Rsetup is used now to set Host or Peripheral mode on the OTG USB3 port. When I look in "/boot/hw_intfc.conf" I see this:
![image](https://github.com/marcone/teslausb/assets/50762577/81a73d0e-737c-4af6-9fb7-93b3ca9f6aac)

I used rsetup utility to set the port to peripheral mode but it enabled an Eth-over-USB function. I removed that using the command "systemctl disable radxa-adbd radxa-usbnet" which fixed that problem but I still can't get it to be detected as a mass storage device in Windows.

At this point I am stuck. I am sure this can work but something about the overlays is not set correctly and I don't know enough about them to modify the state. I can't get the Bookworm release from Armian to boot but I think that's because I have my Rock Pi 4C+ connected to a 4K display. I have a lower-resolution display coming today to try that out.

What revision is your Rock Pi E board and how about that guide?

**Reply by dynm** [reply]
I am using a Rock Pi E v1.2 with Armbian 23.8.3 Bookworm installed on the eMMC module, and the other steps are consistent with the original post.

**Comment by natarajgithub** [comment]
just want to put a quick update. The same works in Rock Pi E v3.0 with Armbian 25.2.3

Was also able to integrade [spotDL](https://github.com/spotDL/spotify-downloader) with minor change in copy-music.sh
This makes the Teslausb to download and update predefined spotify playlists when archive loop is running at home
